Unit 2.3 LangGraph : Release 11/03/25 (#271)
huggingface/agents-course: units/en/unit2/langgraph/when_to_use_langgraph.mdx

> ## What is `LangGraph`?

Reviewer: ```suggestion
# What is `LangGraph`?
```

> `LangGraph` is a framework developed by [LangChain](https://www.langchain.com/) to manage the control flow of applications that integrate an LLM.

Reviewer 2: ```suggestion
`LangGraph` is a framework developed by [LangChain](https://www.langchain.com/) **to manage the control flow of applications that integrate an LLM**.
```

> ## Is `LangGraph` different from `LangChain` ?

Reviewer: ```suggestion
## Is `LangGraph` different from `LangChain`?
```

LangChain provides a standard interface to interact with models and other components, useful for retrieval, LLM calls and tools calls.

> The classes from LangChain might be used in LangGraph, but does not HAVE to be used. 

Reviewer: ```suggestion
The classes from LangChain might be used in LangGraph, but do not HAVE to be used. 
```

> The classes from LangChain might be used in LangGraph, but does not HAVE to be use. The packages are different.

Reviewer 2: I don't understand this phrase. WDYM?

The packages are different and can be used in isolation, but, in the end, all ressources you will find online use both packages hand in hand.

> ## When should I use `LangGraph` ?

Reviewer: ```suggestion
## When should I use `LangGraph`?
```

### Control vs freedom

When designing AI applications, you face a fundamental trade-off between **control** and **freedom**:

- **Freedom** gives your LLM more room to be creative and tackle unexpected problems.
- **Control** allows you to ensure predictable behavior and maintain guardrails.

> Code Agents like the ones you can encounter in smolagents are very free, they can call multiples tools in a single action step, they can create their own tools, etc ... But this behaviour can make them less predictable and less controlable than a regular Agent working with JSON !

Reviewer 2: ```suggestion
Code Agents like the ones you can encounter in *smolagents* are very free, they can call multiples tools in a single action step, they can create their own tools, etc ... But this behaviour can make them less predictable and less controlable than a regular Agent working with JSON !
```

> Code Agents like the ones you can encounter in *smolagents* are very free, they can call multiples tools in a single action step, they can create their own tools, etc ... But this behaviour can make them less predictable and less controlable than a regular Agent working with JSON !

Reviewer: ```suggestion
Code Agents, like the ones you can encounter in *smolagents*, are very free. They can call multiple tools in a single action step, create their own tools, etc. However, this behavior can make them less predictable and less controllable than a regular Agent working with JSON!
```

> `LangGraph` is on the other end of the spectrum, it shines when you need **"Control"** on the execution of your agent. 

Reviewer 2: Les deux phrases sont redondantes

LangGraph is particularly valuable when you need **Control over your agent applications**. It gives you the tools to build an agent application that follows a predictable process while still leveraging the power of LLMs. 

> Put simply, if your application involves a series of steps that need to be orchestrated in a specific way, with decisions being made at each junction point, LangGraph provides the structure you need.

Reviewer 2: ```suggestion
Put simply, if your application involves a series of steps that need to be orchestrated in a specific way, with decisions being made at each junction point, **LangGraph provides the structure you need**.
```

> As an example, let's say we want to build an LLM assistant that can answer some question over some documents.

Reviewer 2: ```suggestion
As an example, let's say we want to build an LLM assistant that can answer some questions over some documents.
```

- **Workflows that need human-in-the-loop interventions**
- **Complex agent architectures** with multiple components working together

In essence, whenever you can, **as a Human** design a flow of actions that needs to be taken, and that, depending on the output of that one action, decide what you should execute next. Then LangGraph is the correct framework for you !

`LangGraph`, is, in my own opinion the most production ready agent framework on the market.

## How does LangGraph work?

At its core, `LangGraph` uses a directed graph structure to define the flow of your application:

- **Nodes** represent individual processing steps (like calling an LLM, using a tool, or making a decision).
- **Edges** define the possible transitions between steps.
- **State** is user defined and maintained and passed between nodes during execution. When deciding which node to target next, this is the current state that we look at.

We will explore those fundamental blocks more in the next chapter ! 

## How is it different from regular python ? Why do I need LangGraph ?

You might wonder: "I could just write regular Python code with if-else statements to handle all these flows, right?" 

> While technically true, LangGraph offers **significant advantages** over vanilla Python for building complex agent systems. You could build the same application without langGraph, but it builds easier tools and abstractions for you.

Reviewer 2: ```suggestion
While technically true, LangGraph offers **significant advantages** over vanilla Python for building complex agent systems. You could build the same application without LangGraph, but it builds easier tools and abstractions for you.
```
